tfhe/src/integer/server_key/radix_parallel/ilog2.rs

```diff
@@ -0,0 +1,734 @@
+        }
+    }
+}
+
+/// Direction to count consecutive bits
+#[derive(Copy, Clone, Eq, PartialEq)]
+enum Direction {
+    /// Count starting from the LSB
+    Trailing,
+    /// Count starting from MSB
+    Leading,
+}
+
+impl ServerKey {
+    /// This function takes a ciphertext in radix representation
+    /// and returns a vec of blocks, where each blocks holds the number of leading_zeros/ones
+    ///
+    /// This contains the logic of making a block have 0 leading_ones/zeros if its preceding
+    /// block was not full of ones/zeros
+    fn prepare_count_of_consecutive_bits<T>(
+        &self,
+        ct: T,
+        direction: Direction,
+        bit_value: BitValue,
+    ) -> Vec<Ciphertext>
+    where
+        T: IntegerRadixCiphertext,
+    {
+        assert!(self.carry_modulus().0 >= self.message_modulus().0);
+
+        let mut blocks = ct.into_blocks();
+
+        let lut = match direction {
+            Direction::Trailing => self.key.generate_lookup_table(|x| {
+                let x = x % self.key.message_modulus.0 as u64;
+
+                let mut count = 0;
+                for i in 0..self.key.message_modulus.0.ilog2() {
+                    if (x >> i) & 1 == bit_value.opposite() as u64 {
+                        break;
+                    }
+                    count += 1;
+                }
+                count
+            }),
+            Direction::Leading => self.key.generate_lookup_table(|x| {
+                let x = x % self.key.message_modulus.0 as u64;
+
+                let mut count = 0;
+                for i in (0..self.key.message_modulus.0.ilog2()).rev() {
+                    if (x >> i) & 1 == bit_value.opposite() as u64 {
+                        break;
+                    }
+                    count += 1;
+                }
+                count
+            }),
+        };
+
+        // Assign to each block its number of leading/trailing zeros/ones
+        // in the message space
+        blocks.par_iter_mut().for_each(|block| {
+            self.key.apply_lookup_table_assign(block, &lut);
+        });
+
+        if direction == Direction::Leading {
+            // Our blocks are from lsb to msb
+            // `leading` means starting from the msb, so we reverse block
+            // for the cum sum process done later
+            blocks.reverse();
+        }
+
+        // Use hillis-steele cumulative-sum algorithm
+        // Here, each block either keeps his value (the number of leading zeros)
+        // or becomes 0 if the preceding block
+        // had a bit set to one in it (leading_zeros != num bits in message)
+        let num_bits_in_message = self.key.message_modulus.0.ilog2() as u64;
+        let sum_lut = self.key.generate_lookup_table_bivariate(
+            |block_num_leading_zeros, more_significant_block_leading_zeros| {
+                if more_significant_block_leading_zeros == num_bits_in_message {
+                    block_num_leading_zeros
```

Comment:
I'm guessing the name is not entirely accurate now ? leading_bits_of_interest or something ?